src/rulesets/LinearAlgebra/matfun.jl

```diff
@@ -0,0 +1,243 @@
+#####
+
+function frule((_, ΔA), ::typeof(LinearAlgebra.exp!), A::StridedMatrix{<:BlasFloat})
+    if ishermitian(A)
+        hermX, ∂hermX = frule((Zero(), ΔA), exp, Hermitian(A))
+        X = LinearAlgebra.copytri!(parent(hermX), 'U', true)
+        if ∂hermX isa LinearAlgebra.RealHermSymComplexHerm
+            ∂X = LinearAlgebra.copytri!(parent(∂hermX), 'U', true)
+        else
+            ∂X = ∂hermX
+        end
+    else
+        X, intermediates = _matfun!(exp, A)
+        ∂X = _matfun_frechet!(exp, A, X, ΔA, intermediates)
+    end
+    return X, ∂X
+end
+
+function rrule(::typeof(exp), A0::StridedMatrix{<:BlasFloat})
+    # TODO: try to make this more type-stable
+    if ishermitian(A0)
+        # call _matfun instead of the rrule to avoid hermitrizing ∂A in the pullback
+        hermA = Hermitian(A0)
+        hermX, hermX_intermediates = _matfun(exp, hermA)
+        function exp_pullback_hermitian(ΔX)
+            ∂hermA = _matfun_frechet(exp, hermA, hermX, ΔX, hermX_intermediates)
+            ∂hermA isa LinearAlgebra.RealHermSymComplexHerm || return NO_FIELDS, ∂hermA
+            return NO_FIELDS, parent(∂hermA)
+        end
+        return LinearAlgebra.copytri!(parent(hermX), 'U', true), exp_pullback_hermitian
+    else
+        A = copy(A0)
+        X, intermediates = _matfun!(exp, A)
+        function exp_pullback(ΔX)
+            ΔX′ = copy(adjoint(ΔX))
+            ∂A′ = _matfun_frechet!(exp, A, X, ΔX′, intermediates)
+            ∂A = copy(adjoint(∂A′))
+            return NO_FIELDS, ∂A
+        end
+        return X, exp_pullback
+    end
+end
+
+## Destructive matrix exponential using algorithm from Higham, 2008,
+## "Functions of Matrices: Theory and Computation", SIAM
+## Adapted from LinearAlgebra.exp! with return of intermediates
+function _matfun!(::typeof(exp), A::StridedMatrix{T}) where T<:BlasFloat
+    n = LinearAlgebra.checksquare(A)
+    ilo, ihi, scale = LAPACK.gebal!('B', A)    # modifies A
+    nA   = opnorm(A, 1)
+    Inn    = Matrix{T}(I, n, n)
+    ## For sufficiently small nA, use lower order Padé-Approximations
+    if (nA <= 2.1)
+        if nA > 0.95
+            C = T[17643225600.,8821612800.,2075673600.,302702400.,
+                     30270240.,   2162160.,    110880.,     3960.,
+                           90.,         1.]
+        elseif nA > 0.25
+            C = T[17297280.,8648640.,1995840.,277200.,
+                     25200.,   1512.,     56.,     1.]
+        elseif nA > 0.015
+            C = T[30240.,15120.,3360.,
+                    420.,   30.,   1.]
+        else
+            C = T[120.,60.,12.,1.]
+        end
+        si = 0
+    else
+        C = T[64764752532480000.,32382376266240000.,7771770303897600.,
+                1187353796428800.,  129060195264000.,  10559470521600.,
+                    670442572800.,      33522128640.,      1323241920.,
+                        40840800.,           960960.,           16380.,
+                             182.,                1.]
```

Comment:
In generalk I am ok with this formatting, except should space after comma.
It did make me think this was a matrix rather than a vector initially.
But it's fine

```suggestion
            C = T[17643225600., 8821612800., 2075673600., 302702400.,
                     30270240.,    2162160.,     110880.,      3960.,
                           90.,          1.]
        elseif nA > 0.25
            C = T[17297280., 8648640., 1995840., 277200.,
                     25200.,    1512.,      56.,      1.]
        elseif nA > 0.015
            C = T[30240., 15120., 3360.,
                    420.,    30.,    1.]
        else
            C = T[120., 60., 12., 1.]
        end
        si = 0
    else
        C = T[64764752532480000., 32382376266240000., 7771770303897600.,
                1187353796428800.,   129060195264000.,   10559470521600.,
                    670442572800.,       33522128640.,       1323241920.,
                        40840800.,            960960.,            16380.,
                             182.,                 1.]
```

Reply:
fwiw, JuliaFormatter BlueStyle tells me this "should" be
```suggestion
            C = T[
                17643225600.0,
                8821612800.0,
                2075673600.0,
                302702400.0,
                30270240.0,
                2162160.0,
                110880.0,
                3960.0,
                90.0,
                1.0,
            ]
        elseif nA > 0.25
            C = T[17297280.0, 8648640.0, 1995840.0, 277200.0, 25200.0, 1512.0, 56.0, 1.0]
        elseif nA > 0.015
            C = T[30240.0, 15120.0, 3360.0, 420.0, 30.0, 1.0]
        else
            C = T[120.0, 60.0, 12.0, 1.0]
        end
        si = 0
    else
        C = T[
            64764752532480000.0,
            32382376266240000.0,
            7771770303897600.0,
            1187353796428800.0,
            129060195264000.0,
            10559470521600.0,
            670442572800.0,
            33522128640.0,
            1323241920.0,
            40840800.0,
            960960.0,
            16380.0,
            182.0,
            1.0,
        ]
```

Reply:
i think running JuliaFormatter on this file would be good :)